pkg/cluster/client.go

```diff
@@ -180,86 +151,141 @@ func (c *Client) LeaderID() string {
 // ListHosts returns a map of host ids to host structures containing metadata
 // and job lists.
 func (c *Client) ListHosts() ([]host.Host, error) {
-	if c := c.local(); c != nil {
-		return c.ListHosts()
-	}
-	client, err := c.RPCClient()
-	if err != nil {
-		return nil, err
-	}
-	var state []host.Host
-	return state, client.Call("Cluster.ListHosts", struct{}{}, &state)
+	var hosts []host.Host
+	return hosts, c.c.Get("/cluster/hosts", &hosts)
 }
 
 // AddJobs requests the addition of more jobs to the cluster.
-func (c *Client) AddJobs(req *host.AddJobsReq) (*host.AddJobsRes, error) {
-	if c := c.local(); c != nil {
-		return c.AddJobs(req)
-	}
-	client, err := c.RPCClient()
-	if err != nil {
-		return nil, err
-	}
-	var res host.AddJobsRes
-	return &res, client.Call("Cluster.AddJobs", req, &res)
+// jobs is a map of host id -> new jobs. Returns the state of the cluster after
+// the operation.
+func (c *Client) AddJobs(jobs map[string][]*host.Job) (map[string]host.Host, error) {
+	var hosts map[string]host.Host
+	return hosts, c.c.Post(fmt.Sprintf("/cluster/jobs"), jobs, &hosts)
 }
 
 // DialHost dials and returns a host client for the specified host identifier.
 func (c *Client) DialHost(id string) (Host, error) {
-	// TODO: reuse connection if leader id == id
+	// don't lookup addr if leader id == id
+	if c.LeaderID() == id {
+		return NewHostClient(c.c.URL, nil), nil
+	}
+
 	services := c.service.Select(map[string]string{"id": id})
 	if len(services) == 0 {
 		return nil, ErrNoServers
 	}
-	addr := services[0].Addr
-	rc, err := rpcplus.DialHTTPPath("tcp", addr, rpcplus.DefaultRPCPath, c.dial)
-	return NewHostClient(addr, rc, c.dial), err
+	addr := "http://" + services[0].Addr
+	return NewHostClient(addr, nil), nil
```

Comment:
The second argument is a `*http.Client`, but it looks like no one ever sets it anyway, I guess we could remove the second argument, since it defaults to `http.DefaultClient`, wdyt?